Workaround for type name collisions

SRetip opened the pull request:
Problem:
If we try to generate schema from this schema:
```json
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "definitions": {
        "Foo": {
            "properties": {
                "Bar": {
                    "oneOf": [
                        {
                            "$ref": "#/definitions/FooBar"
                        },
                        {
                            "$ref": "#/definitions/Baz"
                        }
                    ]
                }
            },
            "type": "object"
        },
        "FooBar": {
            "type": "string"
        },
        "Baz": {
            "type": "string"
        }
    }
}
```
we get:
```rust
pub struct Baz(pub String);

pub struct Foo {
    pub bar: Option<FooBar>,
}

pub struct FooBar(pub String);

pub enum FooBar {
    FooBar(FooBar),
    Baz(Baz),
}
```
where struct `FooBar` - structure from definition:
```json
  "FooBar": {
            "type": "string"
        },
```
and enum `FooBar` - enum which come from property definition:
```json
 "Bar": {
                    "oneOf": [
                        {
                            "$ref": "#/definitions/FooBar"
                        },
                        {
                            "$ref": "#/definitions/Baz"
                        }
                    ]
                }
```
and as a result, we have a name conflict.

With my solution result will be:
```rust
pub struct Baz(pub String);

pub struct Foo {
    pub bar: Option<FooBarAlias>,
}

pub struct FooBar(pub String);

pub enum FooBarAlias {
    FooBar(FooBar),
    Baz(Baz),
}
```
I'd be happy to have a feedback :)

ahl reviewed (changes requested):
I appreciate that this solves for the case you have in mind, but the solution seems fairly narrow. For example, it applies to `add_ref_types`, but not to `add_type`. I'm concerned that successively suffixing "Alias" isn't going to produce types that are particularly usable.

Comments describing your approach would be helpful. Its placement was a little confusing. For example, we're doing this during the type finalization pass... but I'm not sure this has anything to do with type finalization.

SRetip commented:
If I understand you correctly, I have corrected what was required:
- put this logic in a separate method for `TypeEntry`; 
- added the call of this function to the `TypeSpace::add_type_with_name(...)` method;
- added comments:
```
// I think we should do it this way, because in this place we're iterating over all 
// created types.
```
As for adding "Alias", I honestly can't think of a way to make it more user friendly. But otherwise the code will simply not compile without this. 

P.S. The "Alias" post-fix is generated only when there is a name conflict.

SRetip commented:
Hi @ahl  sorry if I'm distracting you, maybe you didn't notice the notification of my comment on this pool request.
I would be very happy to hear your thoughts and comments on this pr.

ahl commented on typify-impl/src/defaults.rs:
```diff
@@ -374,6 +374,17 @@ impl TypeEntry {
         }
     }
+
+    pub(crate) fn rename(&mut self, new_name: String) {
```
unrelated to the rest of this file

ahl commented on typify-impl/src/lib.rs:
```diff
@@ -205,6 +205,8 @@ pub struct TypeSpace {
 
     cache: SchemaCache,
 
+    names: HashSet<String>,
```
1. BTreeSet
2. all names are already in name_to_id

ahl commented on typify-impl/src/lib.rs:
```diff
@@ -584,6 +588,9 @@ impl TypeSpace {
             let type_id = TypeId(index);
             let mut type_entry = self.id_to_entry.get(&type_id).unwrap().clone();
             type_entry.finalize(self)?;
+            // I think we should do it this, because in this place we're iterating over all
+            // created types.
```
can we validate when we're inserting types into name_to_id

ahl commented on typify-impl/src/type_entry.rs:
```diff
@@ -514,6 +514,21 @@ impl TypeEntry {
+    pub(crate) fn ensure_unique_name(&mut self, type_space: &mut TypeSpace) {
+        if let Some(mut name) = self.name().cloned() {
+            while type_space.names.contains(&name) {
+                name = format!("{name}Alias");
```
Let's make the new type name iterate through integers e.g. `FooBar2`, `FooBar3` rather than `FooBarAlias` and `FooBarAliasAlias`.

ahl reviewed (changes requested):
This violates the internal consistency of the `TypeSpace` object. I left some more specific notes. This change is not a high priority for me, so I want to set expectations for infrequent responses.

ahl commented on typify-impl/src/lib.rs:
```diff
@@ -846,7 +847,8 @@ impl TypeSpace {
                 let type_id = self.assign();
-                self.name_to_id.insert(name.clone(), type_id.clone());
+                ty.ensure_unique_name(self);
+                self.name_to_id.insert(name, type_id.clone());
```
do I understand correctly that we're using the old name that might have been changed?

ahl commented on typify-impl/src/type_entry.rs:
```diff
@@ -514,6 +514,34 @@ impl TypeEntry {
+            | TypeEntryDetails::Newtype(TypeEntryNewtype { name, .. }) => {
+                *name = new_name;
+            }
+            _ => {}
```
is this reachable?

SRetip replied:
In this context yes, because this method is called only if the structure has a name, and the same match is called in the name method.

ahl commented on typify-impl/src/type_entry.rs:
```diff
@@ -514,6 +514,34 @@ impl TypeEntry {
+    // Ensures that the name is unique within the provided `TypeSpace`.
+    //
+    // This function checks if the current object's name already exists in the `TypeSpace`. If it does,
+    // the function modifies the name by appending "Alias" until a unique name is found. The unique name
```
fix comment; wrap to 80 cols

SRetip commented:
> do I understand correctly that we're using the old name that might have been changed?

Yes, my bad, I fix this

ahl reviewed (approved):
This feels like a bit of a bolt-on, but I understand that it solves a problem for you and may solve a similar problem for others. If there are future PRs, I'd encourage self-review prior to moving the PR out of draft. It's often also helpful to open an issue and ask questions there to receive guidance regarding an acceptable solution.
